crates/ruff_formatter/src/builders.rs

```diff
@@ -328,6 +328,12 @@ pub struct SourcePosition(TextSize);
 
 impl<Context> Format<Context> for SourcePosition {
     fn fmt(&self, f: &mut Formatter<Context>) -> FormatResult<()> {
+        if let Some(FormatElement::SourcePosition(last_position)) = f.buffer.elements().last() {
```

Comment:
In many situations, nodes start and end exactly at the same position. For example:

```python
print("a")
```

The Module, the expression statement, and the call expression all start and end at the same position. This check prevents that we push the same source marker multiple times to:

* Reduce the IR size (write less data)
* Reduce the work during printing (less IR to read, also applies to humans debugging the IR)

Reply:
I was confused why you think that the call expression should start and end at the same position but I made you believe this to be the case. 

It's not that the call expression starts and ends at the same position, but that different nodes can start or end at the same positions

```
[
  source_position(0),
  source_position(0),
  source_position(0),
  source_position(0),
  "print",
  source_position(5),
  source_position(5),
  "(",
  group([
    indent([
      soft_line_break,
      group([source_position(6), "\"a\"", source_position(9)])
    ]),
    soft_line_break
  ]),
  ")",
  source_position(10),
  source_position(10),
  source_position(10),
  hard_line_break,
  source_position(10)
]
```

This is the IR generated for the above example **without** this change. There's a lot of duplicated positions. 

* position(0): The module, expression statement, call expression, and the identifier `print` start at position 0
* position(5): The identifier ends at position 5 and the arguments start at position 5
* position(10): The call arguments, call expression, expression statement, and module all end at position 10

Reply:
Oh right, this makes sense, thank you.